src/cryptonote_core/pulse.cpp

```diff
@@ -844,80 +877,57 @@ round_state wait_for_next_block(uint64_t hf16_height, round_context &context, cr
     return round_state::wait_for_next_block;
   }
 
-  //
-  // NOTE: Block Timing
-  //
-  uint64_t const delta_height = context.wait_for_next_block.height - cryptonote::get_block_height(genesis_block);
-#if 1
-  auto genesis_timestamp      = pulse::time_point(std::chrono::seconds(genesis_block.timestamp));
-  pulse::time_point ideal_timestamp = genesis_timestamp + (TARGET_BLOCK_TIME * delta_height);
-  pulse::time_point prev_timestamp  = pulse::time_point(std::chrono::seconds(top_block.timestamp));
-  context.wait_for_next_block.round_0_start_time =
-      std::clamp(ideal_timestamp,
-                 prev_timestamp + service_nodes::PULSE_MIN_TARGET_BLOCK_TIME,
-                 prev_timestamp + service_nodes::PULSE_MAX_TARGET_BLOCK_TIME);
-#else // NOTE: Debug, make next block start relatively soon
-  pulse::time_point prev_timestamp               = pulse::time_point(std::chrono::seconds(top_block.timestamp));
-  context.wait_for_next_block.round_0_start_time = prev_timestamp + service_nodes::PULSE_ROUND_TIME;
-#endif
-
-  context.wait_for_next_block.height              = curr_height;
-  context.wait_for_next_block.top_hash            = top_hash;
-  context.wait_for_next_block.top_block_timestamp = top_block.timestamp;
-  context.prepare_for_round                       = {};
+  context.wait_for_next_block.round_0_start_time = times.r0_timestamp;
+  context.wait_for_next_block.height             = curr_height;
+  context.wait_for_next_block.top_hash           = times.prev_hash;
+  context.prepare_for_round                      = {};
 
   return round_state::prepare_for_round;
 }
 
 round_state prepare_for_round(round_context &context, service_nodes::service_node_keys const &key, cryptonote::Blockchain const &blockchain)
 {
-
   // Clear memory
   {
     context.transient = {};
     cryptonote::pulse_random_value &old_random_value = context.transient.random_value.send.data;
     auto &old_random_values_array                    = context.transient.random_value.wait.data;
     memwipe(old_random_value.data, sizeof(old_random_value));
     memwipe(old_random_values_array.data(), old_random_values_array.size() * sizeof(old_random_values_array[0]));
+
+    // Store values
+    bool queue_for_round = context.prepare_for_round.queue_for_next_round;
+    uint8_t round        = context.prepare_for_round.round;
+
+    // Blanket clear
+    context.prepare_for_round = {};
+
+    // Restore values
+    context.prepare_for_round.round                = round;
+    context.prepare_for_round.queue_for_next_round = queue_for_round;
   }
 
   if (context.prepare_for_round.queue_for_next_round)
   {
     // Set when an intermediate Pulse stage has failed and we wait on the
     // next round to occur.
     context.prepare_for_round.queue_for_next_round = false;
-    context.prepare_for_round.round++; //TODO: Overflow check
+
+    if (static_cast<int>(context.prepare_for_round.round + 1) > 255)
```

Comment:
The `static_cast` is unnecessary here: `uint8_t + int` already has to be promoted to int.  (And if the static_cast was necessary, it would have to be on just the first value because otherwise the +1 would have already overflowed it).

Edit: but also simpler is to just drop the +1: `if (prepare_for_round.round >= 255)`.  The `>=` instead of `==` is sort of pedantic but gets the point across.

Edit 2: or alternatively: `if (prepare_for_round.round + uint8_t{1} == 0) // would overflow`

Reply:
Went for  `prepare_for_round.round >= 255` for being the least wordy